### DISCUSSION TITLE
[FastData] Necessary indexes on projection changes collection

### DISCUSSION THREAD

**Discussion by malta895** [mia-platform/community#32] [Q&A] [answered]
Hello, 
in a fast data instance we are maintaining we noticed a lot of duplicates on the projection changes collection, on the `type` and `identifier.<subfield>` fields.

Is there a unique index we can use to avoid this undesired behavior?
Are there any other indexes suggested for this collection?

I couldn't find the answer on Mia-Platform documentation.

Thanks!

**Comment by rstagi** [comment]
Hi,
You're right about the documentation, we'll take care of that!

The index to avoid such behaviour should created by default by the Console. Are you using a custom CRUD Collection?

However, in order to avoid duplicates on your projection changes, the index you're looking for needs to be a **unique index** with the following fields:
```
{ type: 1, identifier: 1 }
```

You can create it directly from the CRUD Collection page using the Indexes, as [explained in the Docs](https://docs.mia-platform.eu/docs/runtime_suite/crud-service/overview_and_usage#indexes).

Hope that helps!

**Reply by malta895** [reply, marked as answer]
Hi! Thanks for your answer! I don't know why the index is missing, it has probably been deleted by mistake.

We deleted the duplicates and created the index as you suggested, and this fixed the problem.

**Comment by danibix95** [comment]
Hi, does your projection changes' `identifier` document, that is your Single View key, contains more than one field?

If that is the case, could you please verify whether duplicated projection changes records have the fields of `identifier` embedded document expressed in different order?

In MongoDB, an equality condition between two embedded documents is matched when:
- each field has the same value
- both embedded document have the same fields
- the order in which fields of embedded documents appear is the same (fields order matters)

[Here](https://www.mongodb.com/docs/manual/reference/method/db.collection.find/#query-embedded-documents) is provided the MongoDB documentation explaining the above concept.

Consequently, the unique index that is currently automatically generated for the projection-changes collection
```json
{ "type": 1, "identifier": 1 }
```
might not be sufficient to distinguish projection changes records whose identifier contains more than one field.

For example, the following documents would be considered **different** by the defined index:

```json
{ "type": "team", "identifier": { "id": 42, "name": "tuna" } }
```
```json
{ "type": "team", "identifier": { "name": "tuna", "id": 42 } }
```
and hence they would appear *"duplicated"* in the collection according to Fast Data logic.

In order to solve this situation it is be necessary to creare an unique index on the specific fields of `identifier` embedded document, so that only fields' values are taken into account.

E.g.
```json
{ "type": "team", "identifier.id": 1, "identifier.name": 1 }
```

As a final note, please remember that the two indexes support different types of queries. For more details you could have a look at the following resources:
- https://www.mongodb.com/docs/manual/core/index-single/#create-an-index-on-an-embedded-field
- https://www.mongodb.com/docs/manual/core/index-single/#create-an-index-on-embedded-document

**Reply by malta895** [reply]
Hi! Thanks for your answer!

In our case the identifier has only one field, but we have different single views with different identifier field names. For this reason, in our projection changes collection we have a situation like the following:
```json
{ "type": "sv1", "identifier": { "sv1Id": "some-id" } }
{ "type": "sv2", "identifier": { "sv2Id": "some-id" } }
{ "type": "sv2", "identifier": { "sv2Id": "some-other-id" } }
```
In this case, following your suggestion, we would have to create  two unique indexes: 
```json
{ "type": 1, "identifier.sv1Id": 1 }
{ "type": 1, "identifier.sv2Id": 1 }
```
But this would lead to a unique constraint violation, since the collection can contain more documents with `identifier.<field-name>: null`, in the example above the second and the third document have `identifier.sv1Id: null`. 

A solution we found to avoid this problem is to define such indexes as partial unique indexes, with the following partial filter expression: 
```json
{
    "identifier.<field-name>" : {
        "$type" : "string"
    }
}
```
The type have to match the one of the identifier field, in our case we have only strings. The null values are excluded by this filter and they are not indexed, since `null` is not of string type.
